Add support for _BitInt on clang

Arghnews opened the pull request:
Issue #4007

Only enabled for clang >= 14 as that's all that supports `_BitInt` at the moment

We need a forwarding reference for `map` to be a "more specialized" match; `const T&`, `T` etc. are ambiguous.

I'm not too familiar, would there be a better way to do this by adding an entry into `enum class type` for `_BitInt`s? Or this is fine?

---

Fyi this worked in version 9 because of this catch-all overload:

https://github.com/fmtlib/fmt/blob/a33701196adfad74917046096bf5a2aa0ab0bb50/include/fmt/core.h#L1506-L1516

Which was then subsequently narrowed down to extended FP types (which "broke" the accidental support for this):

https://github.com/fmtlib/fmt/blob/6a192f8d34cbca3bf6a06575d8f159043d8c7f79/include/fmt/base.h#L1527-L1537

phprus commented on include/fmt/base.h:
```diff
@@ -474,22 +482,23 @@ FMT_CONSTEXPR auto compare(const Char* s1, const Char* s2, std::size_t n)
 
-namespace adl {
-using namespace std;
+namespace detect {
```
Broken rebase

Arghnews replied:
My bad, should be fixed now

vitaut commented on include/fmt/base.h:
```diff
@@ -427,6 +427,14 @@ enum class uint128_opt {};
 #endif
 
+#ifndef FMT_USE_BITINT
+#  if FMT_CLANG_VERSION && FMT_CLANG_VERSION >= 1400
```
This can be simplified to 

```
#  if FMT_CLANG_VERSION >= 1400
```

vitaut commented on include/fmt/base.h:
```diff
@@ -1484,6 +1492,20 @@ template <typename Context> struct arg_mapper {
   FMT_MAP_API auto map(double val) -> double { return val; }
   FMT_MAP_API auto map(long double val) -> long double { return val; }
 
+#if FMT_USE_BITINT
+#  pragma clang diagnostic push
+#  pragma clang diagnostic ignored "-Wbit-int-extension"
+  template <class T, int N = 0> struct is_bitint : std::false_type {};
+  template <int N> struct is_bitint<_BitInt(N)> : std::true_type {};
+  template <int N> struct is_bitint<unsigned _BitInt(N)> : std::true_type {};
```
Let's move this out of the mapper, next to the `_BitInt` detection above.

vitaut commented on include/fmt/base.h:
```diff
@@ -1484,6 +1492,20 @@ template <typename Context> struct arg_mapper {
+#if FMT_USE_BITINT
+#  pragma clang diagnostic push
+#  pragma clang diagnostic ignored "-Wbit-int-extension"
+  template <class T, int N = 0> struct is_bitint : std::false_type {};
```
This should be defined outside of conditional block so that the check could be used on all platforms.

vitaut commented on test/format-test.cc:
```diff
@@ -2442,3 +2442,30 @@ FMT_END_NAMESPACE
+using BitIntS7 = _BitInt(7);
+using BitIntU32 = unsigned _BitInt(32);
+FMT_BEGIN_NAMESPACE
+template <> struct formatter<BitIntS7> : formatter<int8_t> {};
+
+template <> struct formatter<BitIntU32> : formatter<uint32_t> {};
```
Defining formatters for `_BitInt` should be disallowed.

vitaut commented on include/fmt/base.h:
```diff
@@ -1484,6 +1492,20 @@ template <typename Context> struct arg_mapper {
+  template <int N> struct is_bitint<unsigned _BitInt(N)> : std::true_type {};
+
+  template <class T, FMT_ENABLE_IF(is_bitint<remove_cvref_t<T>>::value)>
+  FMT_MAP_API auto map(T&& val) -> decltype(val) {
```
`_BitInt` should be converted to appropriate integer type.

Arghnews replied:
Are we assuming the maximum width for a `_BitInt` is 128 bits here, so we can widen to any of the native integer types (including `__int128`)?

I think [officially](https://releases.llvm.org/14.0.0/tools/clang/docs/ReleaseNotes.html#c-language-changes-in-clang) this is what clang supports according to `BITINT_MAXWIDTH` for c++

Although for C(23), it supports up to 8388608
https://godbolt.org/z/jzfGYoY6a

vitaut replied:
I think we can assume that for now and in any case for smaller sizes we should convert to integers for efficiency instead of going through the formatter extension.

vitaut commented:
@Arghnews, do you plan to update this PR or shall we close it for now (can be reopened later)?

Arghnews commented:
> @Arghnews, do you plan to update this PR or shall we close it for now (can be reopened later)?

@vitaut  Hi, sorry been busy and now I'm on holiday for 2 weeks, but will update once back and implement your feedback. I'd say leave it open and I'll update in 2-3 weeks, but it's up to you (can reopen of course). Cheers!

vitaut commented:
No hurry. Let's keep it open and enjoy your holiday!

Arghnews commented:
Hi,

Reworked this so that now fmtlib will format `_BitInt`s up to 128 bits

Uses wider formatter ie. integer types are widened to 64bit if <= 64 bits wide (assume this is acceptable for performance)
If `FMT_USE_INT128` then supports up to 128 bits

Can't write `static_assert(fmt::is_formattable<signed_bitint<129>, char>{}, "");` as apple clang only allows up to 128 bits. `BITINT_MAXWIDTH` isn't defined, could check `__apple_build_version__` but I just removed this test

Let me know any feedback @vitaut

vitaut commented on include/fmt/base.h:
```diff
@@ -1475,6 +1502,19 @@ template <typename Context> struct arg_mapper {
   FMT_MAP_API auto map(double val) -> double { return val; }
   FMT_MAP_API auto map(long double val) -> long double { return val; }
 
+#if FMT_USE_BITINT
```
Let's remove this check and make sure that the overloads below are disabled if _BitInt is not available.

vitaut commented on include/fmt/base.h:
```diff
@@ -427,6 +427,33 @@ enum class uint128_opt {};
+#  endif
+#endif
+
+template <class T, int N = 0> struct bitint_traits : std::false_type {};
```
I don't think this should inherit from `false_type`. Instead you might need to add `is_formattable = false` to disable overloads for types other than `_BitInt`.

vitaut commented on include/fmt/base.h:
```diff
@@ -427,6 +427,33 @@ enum class uint128_opt {};
+
+// fmt only supports up to 128 bits https://github.com/fmtlib/fmt/pull/4072
+template <int N> struct bitint_traits<_BitInt(N)> : std::true_type {
+  static constexpr bool is_signed = true;
+  static constexpr size_t number_of_bits = N;
```
Can we use numeric_limits instead?

Arghnews replied:
> Can we use numeric_limits instead?

Not if we want this to work when using clang with libstdc++. The libstdc++ implementation just has a load of specialisations for deciding what is signed, and `_BitInt` isn't in it, so it doesn't behave correctly - see https://godbolt.org/z/86hsr5dqK

This works fine with libc++ (which uses an internal function)

Thus I'd say best to leave this as is

vitaut commented on include/fmt/base.h:
```diff
@@ -3157,6 +3197,9 @@ FMT_INLINE void println(format_string<T...> fmt, T&&... args) {
 }
 
 FMT_END_EXPORT
+#if FMT_USE_BITINT
+#  pragma clang diagnostic pop
+#endif
```
Let's move this to the end of the bitint_traits block above.

vitaut commented on include/fmt/format.h:
```diff
@@ -3950,6 +3950,26 @@ class formatter<std::basic_string<Char, Traits, Allocator>, Char>
 template <typename Char, size_t N>
 struct formatter<Char[N], Char> : formatter<basic_string_view<Char>, Char> {};
 
+#if FMT_USE_BITINT
```
Let's use SFINAE instead of conditional compilation.

Arghnews replied:
We probably could, but for example on gcc with `pedantic` it will error with:

```
include/fmt/format.h:3962:46: error: ISO C++ does not support ‘__int128’ for ‘type name’ [-Wpedantic]
 3962 |                     conditional_t<is_signed, __int128, unsigned __int128>,
```

I can disable the warning for that block etc. but not sure it's worthit, especially if msvc and other platforms may have other errors

vitaut commented on include/fmt/format.h:
```diff
@@ -3950,6 +3950,26 @@ class formatter<std::basic_string<Char, Traits, Allocator>, Char>
+};
+}  // namespace detail
+
+template <typename T, typename Char>
+struct formatter<T, Char,
+                 enable_if_t<(detail::bitint_traits<T>::is_formattable)>>
+    : formatter<typename detail::bit_int_format_as<T>::type, Char> {};
```
We could probably use `format_as` instead of specialization.

Arghnews replied:
> We could probably use `format_as` instead of specialization.

https://github.com/fmtlib/fmt/blob/a3f3f2ec9a2657acf0b5a762dc3979ac4900b9d7/include/fmt/base.h#L1420-L1428

Unless I'm misunderstanding what you mean, `format_as` will only work on `enum` or `class` types, of which `_BitInt` is neither?

vitaut replied:
You are absolutely right.

vitaut commented on test/format-test.cc:
```diff
@@ -2507,3 +2509,67 @@ TEST(format_test, writer) {
 }
+
+#if FMT_USE_BITINT
+#  pragma clang diagnostic push
```
We don't need to restore warnings with push/pop in tests.

vitaut commented on test/format-test.cc:
```diff
@@ -2507,3 +2509,67 @@ TEST(format_test, writer) {
+
+#  if FMT_USE_BITINT
+TEST(format_test, bitint) {
+  EXPECT_EQ(fmt::format("{}", unsigned_bitint<3>{7}), "7");
```
We only use braced initialization for aggregates in {fmt}: `unsigned_bitint<3>{7} -> unsigned_bitint<3>(7)`.

vitaut reviewed (changes requested):
A few more comments.

vitaut commented on test/format-test.cc:
```diff
@@ -2507,3 +2509,67 @@ TEST(format_test, writer) {
+  EXPECT_EQ(fmt::format("{}", signed_bitint<128>(0)), "0");
+  EXPECT_EQ(fmt::format("{}", unsigned_bitint<128>(0)), "0");
+  EXPECT_EQ("9223372036854775808",
+            fmt::format("{}", signed_bitint<65>(INT64_MAX) + 1));
```
Let's use `numeric_limits` instead of C macros.

Arghnews replied:
> Let's use `numeric_limits` instead of C macros.

Sure. I assumed you'd be fine with these since I just copied them from higher up in this file

https://github.com/fmtlib/fmt/blob/e3676ca309fd6c128250f65cd4d32bd9f22e2ef9/test/format-test.cc#L1254-L1262

vitaut replied:
If we already use macros elsewhere then it's fine to keep them here.

vitaut commented on include/fmt/format.h:
```diff
@@ -3950,6 +3950,26 @@ class formatter<std::basic_string<Char, Traits, Allocator>, Char>
+  using traits = bitint_traits<T>;
+  static constexpr size_t N = traits::number_of_bits;
+  static constexpr bool is_signed = traits::is_signed;
+  using type = conditional_t<
+      N <= 64, conditional_t<is_signed, int64_t, uint64_t>,
+      conditional_t<(N > 64 && N <= 128),
+                    conditional_t<is_signed, __int128, unsigned __int128>,
+                    void>>;
```
I think this can be moved to `bitint_traits` which would make `number_of_bits` and `is_signed` unnecessary.

Arghnews replied:
> I think this can be moved to `bitint_traits` which would make `number_of_bits` and `is_signed` unnecessary.

Nice suggestion, have done. Only thing to note, we must `#include <cstdint>` in `base.h` now, for the `int64_t`, if this is acceptable

vitaut replied:
Let's use `(unsigned) long long` which is guaranteed to be at least 64-bit and avoid additional include.

Arghnews replied:
> Let's use `(unsigned) long long` which is guaranteed to be at least 64-bit and avoid additional include.

Good shout, have made this change

vitaut commented:
Merged, thanks!
